### PR TITLE
[SPARK-12865][SPARK-12866][SQL] Migrate SparkSQLParser/ExtendedHiveQlParser commands to new Parser

### PR DESCRIPTION
**Pull request by hvanhovell**
This PR moves all the functionality provided by the SparkSQLParser/ExtendedHiveQlParser to the new Parser hierarchy (SparkQl/HiveQl). This also improves the current SET command parsing: the current implementation swallows `set role ...` and `set autocommit ...` commands, this PR respects these commands (and passes them on to Hive).

This PR and https://github.com/apache/spark/pull/10723 end the use of Parser-Combinator parsers for SQL parsing. As a result we can also remove the `AbstractSQLParser` in Catalyst.

The PR is marked WIP as long as it doesn't pass all tests.

cc @rxin @viirya @winningsix (this touches https://github.com/apache/spark/pull/10144)

### CODE REVIEW

**Comment by rxin** [comment]
"The PR is marked WIP as long as it doesn't pass all tests."

This is obvious so if you are only looking to run through the tests, we can just remove the WIP now.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ASTNode.scala [hunk 1]
```diff
@@ -60,6 +60,12 @@ case class ASTNode(
   /** Source text. */
   lazy val source = stream.toString(startIndex, stopIndex)
 
+  /** Get the source text that remains after this token. */
+  lazy val remainder = {
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ASTNode.scala]: if you are updating the pr, can you add explicit types for all the public vals?

File: sql/catalyst/src/main/antlr3/org/apache/spark/sql/catalyst/parser/SparkSqlParser.g [hunk 2]
```diff
@@ -2438,12 +2455,11 @@ BEGIN user defined transaction boundaries; follows SQL 2003 standard exactly exc
 sqlTransactionStatement
 @init { pushMsg("transaction statement", state); }
 @after { popMsg(state); }
-  :
-  startTransactionStatement
-	|	commitStatement
-	|	rollbackStatement
-	| setAutoCommitStatement
-	;
```
Comment (hvanhovell) [inline sql/catalyst/src/main/antlr3/org/apache/spark/sql/catalyst/parser/SparkSqlParser.g]: Removed TAB's...

File: sql/catalyst/src/main/antlr3/org/apache/spark/sql/catalyst/parser/SparkSqlParser.g [hunk 3]
```diff
@@ -687,8 +699,12 @@ catch (RecognitionException e) {
 
 // starting rule
 statement
-	: explainStatement EOF
-	| execStatement EOF
+    : explainStatement EOF
+    | execStatement EOF
+    | KW_ADD KW_JAR -> ^(TOK_ADDJAR)
```
Comment (hvanhovell) [inline sql/catalyst/src/main/antlr3/org/apache/spark/sql/catalyst/parser/SparkSqlParser.g]: These match rules do not match until EOF. We will use the raw input after we encouter such a token.

File: sql/catalyst/src/main/antlr3/org/apache/spark/sql/catalyst/parser/ExpressionParser.g [hunk 4]
```diff
@@ -553,6 +551,8 @@ nonReserved
     | KW_SNAPSHOT
     | KW_AUTOCOMMIT
     | KW_ANTI
+    | KW_WEEK | KW_MILLISECOND | KW_MICROSECOND
```
Comment (hvanhovell) [inline sql/catalyst/src/main/antlr3/org/apache/spark/sql/catalyst/parser/ExpressionParser.g]: This is a leftover chore from a previous PR.

File: sql/catalyst/src/main/antlr3/org/apache/spark/sql/catalyst/parser/ExpressionParser.g [hunk 5]
```diff
@@ -167,8 +167,8 @@ intervalLiteral
       ((intervalConstant KW_HOUR)=> hour=intervalConstant KW_HOUR)?
       ((intervalConstant KW_MINUTE)=> minute=intervalConstant KW_MINUTE)?
       ((intervalConstant KW_SECOND)=> second=intervalConstant KW_SECOND)?
-      (millisecond=intervalConstant KW_MILLISECOND)?
-      (microsecond=intervalConstant KW_MICROSECOND)?
+      ((intervalConstant KW_MILLISECOND)=> millisecond=intervalConstant KW_MILLISECOND)?
```
Comment (hvanhovell) [inline sql/catalyst/src/main/antlr3/org/apache/spark/sql/catalyst/parser/ExpressionParser.g]: This is a leftover chore from a previous PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ASTNode.scala [hunk 6]
```diff
@@ -55,10 +55,16 @@ case class ASTNode(
...
 
   /** Origin of the ASTNode. */
-  override val origin = Origin(Some(line), Some(positionInLine))
+  override val origin: Origin = Origin(Some(line), Some(positionInLine))
 
   /** Source text. */
-  lazy val source = stream.toString(startIndex, stopIndex)
+  lazy val source: String = stream.toString(startIndex, stopIndex)
+
+  /** Get the source text that remains after this token. */
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ASTNode.scala]: would be great to briefly mention what this is used for (set?)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ASTNode.scala [hunk 7]
```diff
@@ -55,10 +55,16 @@ case class ASTNode(
   }
 
   /** Origin of the ASTNode. */
-  override val origin = Origin(Some(line), Some(positionInLine))
+  override val origin: Origin = Origin(Some(line), Some(positionInLine))
 
   /** Source text. */
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ASTNode.scala]: this comment isn't super useful - can you explain a bit more what "source text" is?

**Comment by rxin** [comment]
I'm going to merge this. Please address the minor feedback in a follow-up pr.

**Pull request closed without merging**